src/Control/Monad/IO/Class/Linear.hs

```diff
@@ -0,0 +1,21 @@
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+module Control.Monad.IO.Class.Linear where
+
+import qualified Control.Monad.IO.Class as NonLinear
+import Control.Monad.Lift.Linear
+import Prelude.Linear
+
+-- | Like 'NonLinear.MonadIO' but allows to lift non-linear 'IO'
+-- actions into a linear monad
+type MonadIO m = (MonadLift m, NonLinear.MonadIO (LiftedM m))
```

Comment:
This doesn't feel right. What we want of a linear `MonadIO` is to be able to `liftIO` linear `IO` functions. Not non-linear ones.

If we can lift linear functions, we can in particular lift non-linear `IO` functions (via the `MonadLift IO` instance). We may or may not want to have a corresponding non-iinear monad to lift from. But it should be an orthogonal concern.

Reply:
Let me see if I understand your proposal. `MonadLift` is ok as it is, but MonadIO should be made a type class that allows lifting Linear.IO into a linear monad.

```Haskell
class Linear.Monad m => MonadIO m where
  liftIO :: Linear.IO a ->. m a
```

That seems to work.

Reply:
Indeed, but if we do that, I'm not sure we need `MonadLift` as an abstraction anymore.

Reply:
`directory-server` is using `MonadLift` to run a MonadLogger. I don't think that would be possible with `MonadIO` alone.